src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/Algebras/BDDAlgebra.cs

```diff
@@ -497,30 +507,38 @@ private BDD Shift_rec(Dictionary<ShiftOpKey, BDD> shiftCache, BDD set, int k)
         /// <param name="maxBit">bits above maxBit are unspecified</param>
         public BDD CreateSetFromRange(uint m, uint n, int maxBit)
         {
-            lock (this)
-            {
-                if (n < m)
-                    return False;
 
-                uint mask = (uint)1 << maxBit;
+            if (n < m)
+                return False;
 
-                //filter out bits greater than maxBit
-                if (maxBit < 31)
-                {
-                    uint filter = (mask << 1) - 1;
-                    m &= filter;
-                    n &= filter;
-                }
+            uint mask = (uint)1 << maxBit;
 
+            //filter out bits greater than maxBit
+            if (maxBit < 31)
```

Comment:
Unrelated to your changes, but a left-over that I just recalled. There is an implicit assumption that has to do with 31, obviously related to `maxBit` of `uint` but why `<`. Essentially the intervals->BDD algo is missing some intuition (although I wrote it myself 10y back or so, _I forgot_, there are some internal assumptions, e.g. that it works correctly for characters, i.e., ushort, but actually I don't think it works correctly for arbitrary `uint` intervals due to some overflow) . So definitely some `Debug.Assert`'s are missing (something todo later).

Reply:
I created a todo on our project board for this.